sound/soc/sof/ipc4-topology.c

```diff
@@ -1431,13 +1431,64 @@ static int sof_ipc4_control_load_volume(struct snd_sof_dev *sdev, struct snd_sof
 	return 0;
 }
 
+static int sof_ipc4_control_load_bytes(struct snd_sof_dev *sdev, struct snd_sof_control *scontrol)
+{
+	struct sof_ipc4_control_data *control_data;
+	int ret;
+
+	scontrol->size = sizeof(struct sof_ipc4_control_data) + scontrol->priv_size;
+
+	scontrol->ipc_control_data = kzalloc(scontrol->size, GFP_KERNEL);
+	if (!scontrol->ipc_control_data)
+		return -ENOMEM;
+
+	control_data = scontrol->ipc_control_data;
+	control_data->index = scontrol->index;
+	if (scontrol->priv_size > 0) {
+		memcpy(control_data->data, scontrol->priv, scontrol->priv_size);
+
+		if (control_data->data->magic != SOF_ABI_MAGIC) {
+			dev_err(sdev->dev, "Wrong ABI magic 0x%08x.\n", control_data->data->magic);
+			ret = -EINVAL;
+			goto err;
+		}
+		if (SOF_ABI_VERSION_INCOMPATIBLE(SOF_ABI_VERSION, control_data->data->abi)) {
```

Comment:
hmm. Good question. I never realize this issue. However, such ABI information comes from topology, not firmware. So it can work both for SOF and ACE. And such ABI header is generate by sof-ctl tool. My queston is: is the ABI for FW or topology? If the ABI information means FW, it won't work. If the ABI is for topology, it works.

Reply:
so what are we checking here? kernel ABI and topology ABI? where do we set the kernel ABI?

Reply:
@libinyang can we leave out this check and add a TODO for now. I m not sure how this is supposed to work yet. We need some more discussions around it